### PR TITLE
Add e2e test

### LINKED ISSUE #23: Add k8s matrix test
## What

Add k8s matrix test like TopoLVM using kind.

## How

- Run test using kind or minikube with selecting kubernetes versions.
- Install TopoLVM as a CSI Driver and use it for StorageClass.
- Install prometheus or compatible service for metrics server.
- Enable matrix test on github actions.

## Checklist

- [ ] Finish implementation of the issue
- [ ] Test all functions
- [ ] Have enough logs to trace activities
- [ ] Notify developers of necessary actions

### PR DESCRIPTION
**Pull request by bells17**
resolve #23

I added k8s matrix e2e tests.

### CODE REVIEW

File: e2e/testdata/pod-pvc-template.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: {{ .pvcName }}
+  namespace: {{ .namespace }}
+{{ if eq .useAnnotation "true" }}
+  annotations:
+    resize.topolvm.io/threshold: {{ if ne .thresholdAnnotation "" }}{{ .thresholdAnnotation }}{{ else }}""{{ end }}
```
Comment (toshipp) [inline e2e/testdata/pod-pvc-template.yaml:8]: Can this be `"{{.thresholdAnnotation}}"`?

File: e2e/suite_test.go [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+	"sigs.k8s.io/yaml"
+)
+
+var (
+	binDir string
+
+	//go:embed testdata/pod-pvc-template.yaml
+	podPVCTemplateYAML string
+
+	// proxyConfigOnce guards proxyConfig
```
Comment (toshipp) [inline e2e/suite_test.go:31]: Is it garbage?

**Review by toshipp — CHANGES_REQUESTED** [review]

File: e2e/suite_test.go (hunk 2 above)
Comment (bells17) [reply e2e/suite_test.go]: sorry, I fixed it.

File: e2e/testdata/pod-pvc-template.yaml (hunk 1 above)
Comment (bells17) [reply e2e/testdata/pod-pvc-template.yaml]: Thank you. I fixed it.

File: e2e/autoresizer-cluster.yaml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    extraArgs:
+      config: /var/lib/scheduler/scheduler-config.yaml
+- |
+  apiVersion: "kubeadm.k8s.io/v1beta2"
+  kind: JoinConfiguration
+  metadata:
+    name: config
+  nodeRegistration:
+    kubeletExtraArgs:
+      volume-stats-agg-period: 1s
```
Comment (llamerada-jp) [inline e2e/autoresizer-cluster.yaml:28]: Could you write comments about the difference between this file with TopoLVM one to maintain it?

**Review by toshipp — APPROVED** [review]
LGTM

File: e2e/topolvm-example.mk [hunk 4]
```diff
@@ -0,0 +1,98 @@
+KUBERNETES_VERSION?=1.21.1
+KUSTOMIZE_VERSION=3.8.9
+TOPOLVM_VERSION=0.8.3
+
+KIND_CLUSTER_NAME=autoresizer-e2e
```
Comment (llamerada-jp) [inline e2e/topolvm-example.mk:5]: Would you modify `topolvm/example/Makefile` so that KUBERNETES_VERSION, KIND_CLUSTER_NAME, and TMPDIR can be changed?

Comment (llamerada-jp) [reply e2e/topolvm-example.mk:5]: I made a PR for topolvm/example/Makefile.
https://github.com/topolvm/topolvm/pull/326
Could you check it?

File: .gitignore [hunk 5]
```diff
@@ -2,3 +2,4 @@
 /.vscode
 /.idea
 /testbin
+/vendor
```
[the remark this answers is not in the document]
Comment (bells17) [reply .gitignore:5]: Running the `go mod vendor` command creates a vendor directory.
So I think it's safer to add the vendor directory to the .gitignore file.

File: Makefile [hunk 6]
```diff
@@ -5,6 +5,7 @@ KUBEBUILDER_VERSION = 3.1.0
 CTRLTOOLS_VERSION = 0.6.0
 CTRLRUNTIME_VERSION = 0.8.3
 KUSTOMIZE_VERSION = 3.8.7
+KIND_VERSION=v0.11.1
```
[the remark this answers is not in the document]
Comment (bells17) [reply Makefile]: Sorry, I fixed it.
https://github.com/topolvm/pvc-autoresizer/pull/44/commits/567347a12e0283faa34a245dee66557856279912

File: e2e/.gitignore [hunk 7]
```diff
@@ -0,0 +1,8 @@
+*.img
+*.test
+/build
+/gencert
+/certs
+/kube-prometheus
+/scheduler-config.yaml
+/manifests/topolvm/kustomization.yaml
```
[the remark this answers is not in the document]
Comment (bells17) [reply e2e/.gitignore]: Sorry, I fixed it.
https://github.com/topolvm/pvc-autoresizer/pull/44/commits/567347a12e0283faa34a245dee66557856279912

File: e2e/autoresizer-cluster.yaml (hunk 3 above)
Comment (bells17) [reply e2e/autoresizer-cluster.yaml]: I added comment.
https://github.com/topolvm/pvc-autoresizer/pull/44/commits/cc119abb1acd783869e930e4abd5162cd0e110b3

File: e2e/Makefile [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+		&& mkdir $(shell pwd)/kube-prometheus \
+		&& mv $(shell pwd)/kube-prometheus-$(KUBE_PROMETHEUS_VERSION)/manifests kube-prometheus/manifests \
+		&& rm -fr $(shell pwd)/kube-prometheus-$(KUBE_PROMETHEUS_VERSION)
+
+.PHONY: clean
+clean: stop-lvmd
+	rm -rf $(BINDIR) \
+				 $(TMPDIR) \
+				 autoresizer.img \
+				 kube-prometheus
```
[the remark this answers is not in the document]
Comment (bells17) [reply e2e/Makefile]: Sorry, I fixed it.
https://github.com/topolvm/pvc-autoresizer/pull/44/commits/1280873515a062b8a29af6dde3a2de0cfb278fad

File: .gitignore (hunk 5 above)
[the remark this answers is not in the document]
Comment (llamerada-jp) [reply .gitignore:5]: Sure. We doesn't use vendor, but other user might use it.

File: e2e/topolvm-example.mk (hunk 4 above)
Comment (bells17) [reply e2e/topolvm-example.mk]: Thank you, I fixed Makefile.
https://github.com/topolvm/pvc-autoresizer/pull/44/commits/a72676f71ea34ab13e35334f690b6b31e314c21f

**Review by toshipp — DISMISSED** [review]

File: e2e/autoresizer-cluster.yaml [hunk 9]
```diff
@@ -0,0 +1,43 @@
+apiVersion:  kind.x-k8s.io/v1alpha4
```
Comment (llamerada-jp) [inline e2e/autoresizer-cluster.yaml:1]: Would you please remove this patch if it is not required it? Is it correct that the configuration of `volume-stats-agg-period` is not required to pass the e2e test?

Comment (bells17) [reply e2e/autoresizer-cluster.yaml:1]: Sorry, I fixed it.
https://github.com/topolvm/pvc-autoresizer/pull/44/commits/516f4086639e4183386f279db73517849c4b48be

**Review by toshipp — DISMISSED** [review]

**Comment by bells17** [comment]
I squashed my commits.

**Review by satoru-takeuchi — APPROVED** [review]

**Review by llamerada-jp — APPROVED** [review]
LGTM

**Pull request merged**